File: Day-2/Labs/kafka-dotnet-lab/src/KafkaDotNetLab/SimpleConsumer.cs
using Confluent.Kafka;

namespace KafkaDotNetLab;

internal static class SimpleConsumer
{
    public static void Run(string[] args)
    {
        var bootstrap = KafkaConfig.Arg(args, 0, KafkaConfig.DefaultBootstrap);
        var topic = KafkaConfig.Arg(args, 1, KafkaConfig.DefaultTopic);
        var groupId = KafkaConfig.Arg(args, 2, KafkaConfig.DefaultGroupId);

        Console.WriteLine(
            $"Connecting to {bootstrap}, topic={topic}, group={groupId} (Ctrl+C to stop)");

        var config = new ConsumerConfig
        {
            BootstrapServers = bootstrap,
            GroupId = groupId,
            AutoOffsetReset = AutoOffsetReset.Earliest,
            EnableAutoCommit = true,
            ClientId = "dotnet-lab-consumer",
        };

        using var consumer = new ConsumerBuilder<string, string>(config).Build();
        consumer.Subscribe(topic);

        using var cts = new CancellationTokenSource();
        Console.CancelKeyPress += (_, e) =>
        {
            e.Cancel = true;
            cts.Cancel();
        };

        try
        {
            while (!cts.Token.IsCancellationRequested)
            {
                try
                {
                    var record = consumer.Consume(cts.Token);
                    Console.WriteLine(
                        $"Received: \"{record.Message.Value}\" | " +
                        $"partition={record.Partition.Value} offset={record.Offset.Value} " +
                        $"key={record.Message.Key ?? "null"}");
                }
                catch (ConsumeException ex)
                {
                    Console.Error.WriteLine($"Consume error: {ex.Error.Reason}");
                }
            }
        }
        catch (OperationCanceledException)
        {
            Console.WriteLine("\nConsumer stopped.");
        }
        finally
        {
            consumer.Close();
        }
    }
}

File: Day-2/Labs/kafka-python-lab/simple_consumer.py
"""
Subscribe to a topic and print messages as they arrive.

Run:
    python simple_consumer.py
    python simple_consumer.py localhost:9092 lab-messages my-group

Press Ctrl+C to stop.
"""

import sys

from kafka import KafkaConsumer

from config import DEFAULT_BOOTSTRAP, DEFAULT_GROUP_ID, DEFAULT_TOPIC


def main() -> None:
    bootstrap = sys.argv[1] if len(sys.argv) > 1 else DEFAULT_BOOTSTRAP
    topic = sys.argv[2] if len(sys.argv) > 2 else DEFAULT_TOPIC
    group_id = sys.argv[3] if len(sys.argv) > 3 else DEFAULT_GROUP_ID

    print(
        f"Connecting to {bootstrap}, topic={topic}, group={group_id} (Ctrl+C to stop)"
    )

    consumer = KafkaConsumer(
        topic,
        bootstrap_servers=bootstrap,
        group_id=group_id,
        auto_offset_reset="earliest",
        enable_auto_commit=True,
        value_deserializer=lambda v: v.decode("utf-8"),
        client_id="python-lab-consumer",
    )

    try:
        for message in consumer:
            print(
                f'Received: "{message.value}" | '
                f"partition={message.partition} offset={message.offset} "
                f"key={message.key!r}"
            )
    except KeyboardInterrupt:
        print("\nConsumer stopped.")
    finally:
        consumer.close()


if __name__ == "__main__":
    main()
